# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [1]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric
import random
import pandas as pd

In [2]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [3]:
x.shape

(506, 13)

In [4]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [5]:
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples
    
    input_data = pd.DataFrame(input_data)
    target_data = pd.DataFrame(target_data)
    
    N = len(input_data)
    no_of_samples = int(0.6 * N) #60% of data
    X_indices = input_data.index.tolist()
    
    selected_rows = random.sample(X_indices,no_of_samples)
    selected_columns = np.random.choice(input_data.columns.tolist(),size = random.randint(3,12))
    
    sampled_input_data = input_data.loc[selected_rows,selected_columns]
    
    sampled_target_data = target_data.loc[selected_rows]
    
    #print(len(sampled_input_data))
    #print(sampled_target_data)
    
    #replication
    
    replicated_rows = random.sample(sampled_input_data.index.tolist(),N - no_of_samples)
    
    #Concatenation
    sampled_replicated_rows = sampled_input_data.loc[replicated_rows,:]
    sampled_target_rows = sampled_target_data.loc[replicated_rows,:]
    
    #print(sampled_target_rows)
    
    final_sampled_data = pd.concat([sampled_input_data,sampled_replicated_rows])
    final_target_data = pd.concat([sampled_target_data,sampled_target_rows])
    
    final_sampled_data.reset_index(drop = True,inplace = True)
    final_target_data.reset_index(drop = True,inplace = True)
    
    #print(final_target_data)
    
    return final_sampled_data.values , final_target_data.values,selected_rows,selected_columns
    #note please return as lists

<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [6]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [7]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
list_input_data =[]
list_output_data =[]
list_selected_row= []
list_selected_columns=[]

for i in range(0,30):
    a,b,c,d = generating_samples(x, y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [8]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [9]:
from sklearn.tree import DecisionTreeRegressor
list_of_models = []

for i in range(0,30):
    clf = DecisionTreeRegressor(max_depth=None)
    clf.fit(list_input_data[i],list_output_data[i])
    list_of_models.append(clf)

In [10]:
print(len(list_of_models))
print(type(list_of_models[0]))

30
<class 'sklearn.tree._classes.DecisionTreeRegressor'>


<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [11]:
#a = [1,2,3,4,5,6,7,8,9,10,11,12,13]
#b = pd.DataFrame([np.array(a)])
#b[list_selected_columns[0]].values

In [12]:
from sklearn.metrics import mean_squared_error

def Query(datapoint,selected_columns,models_list):
    datapoint = np.array(datapoint)
    df = pd.DataFrame([datapoint])
    y_pred_temp = []
    for i in range(0,30):
        #datapoint = datapoint.reshape(1,-1)
        
        temp = models_list[i].predict(df[selected_columns[i]].values)
        y_pred_temp.append(temp)
    
    return np.median(np.asarray(y_pred_temp))


def MSE(input_data,output_data,selected_columns,models_list,return_y_pred):
    y_pred = []
    for point in input_data:
        y_pred.append(Query(point,selected_columns,models_list))
    mse = mean_squared_error(output_data,np.array(y_pred))
    
    if return_y_pred == True :
        return mse,y_pred
    
    return mse
    

In [13]:
mse,y_pred = MSE(x,y,list_selected_columns,list_of_models,True)
print("Mean Squared Error",mse)

Mean Squared Error 0.15436649099692581


In [15]:
df1 = pd.DataFrame(
    {
    "Y_True": np.array(y),
    "Y_pred": np.array(y_pred)
    }
)
df1.head()

Y_True  Y_pred
0    24.0    24.0
1    21.6    21.6
2    34.7    34.7
3    33.4    33.4
4    36.2    36.2

In [16]:
a = df1[df1['Y_True'] != df1['Y_pred']]
print(len(a))
a.head()

21


Y_True     Y_pred
68     17.4  18.383333
142    13.4  13.900000
152    15.3  17.000000
165    25.0  24.650000
291    37.3  36.900000

<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [17]:
#a= pd.DataFrame(x)

In [18]:
#a.head()

In [19]:
#b = a[~a.index.isin(list_selected_row[0])]
#len(b)

In [20]:
#b.head()

In [21]:
def Queryoob(datapoint,idx,selected_row,selected_columns,models_list):
    datapoint = np.array(datapoint)
    df = pd.DataFrame([datapoint])
    y_pred_temp = []
    for i in range(0,30):
        #datapoint = datapoint.reshape(1,-1)   
        if idx not in selected_row[i]:
            temp = models_list[i].predict(df[selected_columns[i]].values)
            y_pred_temp.append(temp)
    
    return np.median(np.asarray(y_pred_temp))


def OOB(input_data,output_data,selected_row,selected_columns,models_list,return_y_oob): 
    y_pred_oob = []
    x_train = pd.DataFrame(input_data)
    for idx,point in enumerate(x_train.values):     
        #not_selected_rows = input[~input.index.isin(list_selected_row[i])]
        y_pred_oob.append(Queryoob(point,idx,selected_row,selected_columns,models_list))
    mse = mean_squared_error(output_data,np.array(y_pred_oob))
    #print(len(y_pred_oob))
    
    if return_y_oob == True:
        return mse,y_pred_oob
    return mse

In [22]:
oob,y_pred_oob = OOB(x,y,list_selected_row,list_selected_columns,list_of_models,True)
oob

19.13069466357368

In [23]:
df = pd.DataFrame(
    {
    "Y_True": np.array(y),
    "Y_OOB": np.array(y_pred_oob)
    }
)
df.head()

Y_True  Y_OOB
0    24.0  23.65
1    21.6  25.00
2    34.7  33.20
3    33.4  36.20
4    36.2  30.30

In [24]:
#x_train = pd.DataFrame(x)
#not_selected_rows = x_train[~x_train.index.isin(list_selected_row[0])]
#not_selected_rows.head()

In [25]:
#x_train.values

In [26]:
#len(not_selected_rows)

# <font color='blue'><b>Task 2</b></font>

In [27]:
from tqdm import tqdm

train_mse =  []
oob_mse = []
#y_pred_lists = []
#oob_lists = []


for i in tqdm(range(0,35)):
    input_data_list =[]
    output_data_list =[]
    selected_row_list = []
    selected_columns_list =[]
    models = []
    
    for i in range(0,30):
        a,b,c,d = generating_samples(x, y)
        input_data_list.append(a)
        output_data_list.append(b)
        selected_row_list.append(c)
        selected_columns_list.append(d)
        
    for i in range(0,30):
        clf = DecisionTreeRegressor(max_depth=None)
        clf.fit(input_data_list[i],output_data_list[i])
        models.append(clf)
        
    mse_train = MSE(x,y,selected_columns_list,models,False)
    mse_oob = OOB(x,y,selected_row_list,selected_columns_list,models,False)
    
    train_mse.append(mse_train)
    oob_mse.append(mse_oob)    

100%|██████████████████████████████████████████████████████████████████████████████████| 35/35 [17:23<00:00, 29.82s/it]


In [28]:
print(train_mse)

[0.05797060276679844, 0.08172567832701799, 0.10850790513833992, 0.04687386912604309, 0.0027223320158102794, 0.04968676041819159, 0.13939229249011867, 0.15984807312252963, 0.046973814229248984, 0.18063221343873512, 0.28512448219535824, 0.05015041721563465, 0.08021145589658787, 0.13462185969176133, 0.10497749231444892, 0.034886363636363625, 0.15112242010275795, 0.28726818716226304, 0.27209486166007923, 0.36554668972332033, 0.28525502181419726, 0.012391444446043932, 0.03915911835748791, 0.10467899099692574, 0.05812252964426879, 0.11713897733372367, 0.28264174901185757, 0.32125557090425105, 0.15803194993412384, 0.04219535573122529, 0.06597675120772946, 0.15015851998243304, 0.14302797086962615, 0.20339445761335936, 0.061458745059288576]


In [29]:
print(oob_mse)

[17.882479638779095, 16.81954590791958, 19.533858421168205, 14.509221036061671, 18.031892133763357, 19.937375763408802, 17.72180018267317, 16.881155796585546, 17.37744901942746, 18.134016528559012, 15.151906308223946, 19.285144821663845, 18.57290043202588, 22.23738169337459, 18.691750004097962, 18.39015127497804, 12.14934464589429, 17.045533523243297, 16.771962697628457, 15.556583047410468, 16.609042966321944, 18.301893322919888, 16.94699909797927, 14.791964423320932, 13.497091053194993, 15.41640081837796, 14.533053886869407, 16.69460262122082, 16.163896215305822, 17.904291530327026, 15.414323686746691, 14.585209576374687, 16.495110537982832, 16.897521464012378, 18.06959046991656]


In [30]:
mses = pd.DataFrame(
    {
    "TRAIN_MSE": np.array(train_mse),
    "OOB_MSE": np.array(oob_mse)
    }
)
mses.head()

TRAIN_MSE    OOB_MSE
0   0.057971  17.882480
1   0.081726  16.819546
2   0.108508  19.533858
3   0.046874  14.509221
4   0.002722  18.031892

In [31]:
mses.describe()

TRAIN_MSE    OOB_MSE
count  35.000000  35.000000
mean    0.133864  16.942927
std     0.097952   1.977341
min     0.002722  12.149345
25%     0.054061  15.486492
50%     0.108508  16.897521
75%     0.170240  18.101803
max     0.365547  22.237382

In [32]:
import math
sample_mean_t = np.mean(train_mse)
sample_std_error_t = np.std(train_mse)/math.sqrt(len(train_mse))
print("Confidence Interval Train_MSE 95% Probability")
print(sample_mean_t - (2 * sample_std_error_t),sample_mean_t + (2 * sample_std_error_t))

Confidence Interval Train_MSE 95% Probability
0.10122637218530656 0.16650076630486205


In [33]:
sample_mean_o = np.mean(oob_mse)
sample_std_error_o = np.std(oob_mse)/math.sqrt(len(oob_mse))
print("Confidence Interval OOB_MSE 95% Probability")
print(sample_mean_o - (2 * sample_std_error_o),sample_mean_o + (2 * sample_std_error_o))

Confidence Interval OOB_MSE 95% Probability
16.28408240076639 17.6017715733912


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [34]:
q = [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]
Query(q,list_selected_columns,list_of_models)

18.5

<br><br><br>

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

# Observation:

Task 1 : 

Here we are building multiple number of Decision Trees which has low bias and high variance,
by doing row sampling and column sampling we are keeping bias low and reducing variance.
Row sampling and column sampling controls the overfitting.

In Task 1,mean squared error on training data is very low that means we performing well on training data.
Out of bag points of each model are the point which were not used while training the base learners(DT),
making use of it as cross validation data is good practice as we are not throwing out data.

Task 2 :

95% confidence interval for Train MSE is (0.10122637218530656 0.16650076630486205).That means there is 95% chance that it fall in this range for train_mse.

95% confidence interval for OOB MSE is (16.28408240076639 17.6017715733912).That means there is 95% chance that it fall in this range for obb_mse.

Task 3 :

Here in task 3 we computing predicted value,where we are feeding the input datapoint & it  gives you predicted value(Target).Internally it is calling 30 base learners,based on each base learner which take subset of the features and values. further we are computing median of predicted values returned by each base learners.